core/src/main/scala/algebra/ring/Ring.scala

```diff
@@ -32,11 +32,11 @@ trait Ring[@sp(Int, Long, Float, Double) A] extends Any with Rig[A] with Rng[A]
   def fromInt(n: Int): A = sumN(one, n)
 }
 
-trait RingFunctions {
+trait RingFunctions extends AdditiveGroupFunctions with MultiplicativeMonoidFunctions {
   def fromInt[@sp(Int, Long, Float, Double) A](n: Int)(implicit ev: Ring[A]): A =
     ev.fromInt(n)
 }
 
-object Ring extends AdditiveGroupFunctions with MultiplicativeMonoidFunctions {
+object Ring extends RingFunctions {
   @inline final def apply[A](implicit ev: Ring[A]): Ring[A] = ev
```

Comment:
This is one of those cases where for performance reasons I think using a shared trait is a bad idea -- once you use `Ring.apply`, `CommutativeRing.apply`, and `EuclideanRing.apply` all inlining would be de-optimized (due to the megamorphic call site).

In the long run I would prefer to use *imp* to compile this entire call away -- in that case using a trait would be fine. But until then I'd rather not make this change.